keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,535 @@
+---
+title: ExecutionHook 
+authors:
+  - "@jingxu97"
+  - "@xing-yang"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-storage
+  - sig-node
+  - sig-apps
+  - sig-architecture
+reviewers:
+  - "@saad-ali"
+  - "@thockin"
+  - "@liyinan926"
+approvers:
+  - "@thockin"
+  - "@saad-ali"
+  - "@liyinan926"
+editor: TBD
+creation-date: 2019-1-20
+last-updated: 2019-4-25
+status: implementable
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Title
+
+ExecutionHook API Design
+
+## Table of Contents
+
+  * [Title](#title)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+      * [Workarounds](#workarounds)
+      * [Alternatives](#alternatives)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+
+## Summary
+
+This proposal is to introduce an API (ExecutionHook) for dynamically executing user’s commands in a pod/container or a group of pods/containers and a controller (ExecutionHookController) to manage the hook lifecycle. ExecutionHook provides a general mechanism for users to trigger hook commands in their containers for their different use cases. Different options have been evaluated to decide how this ExecutionHook should be managed and executed. The preferred option is described in the Proposal section. The other options are discussed in the Alternatives section.
```

Comment:
I have a problem with the utilization of the term "hook" in this context. In general, a hook is used "to alter or augment the behaviour of a software system by intercepting function calls or messages or events passed between  components" (adapted from Wikipedia entry). Hooks are related to specific events or actions, not are intended to be arbitrarily called as I understand from the proposal. 

I understand the proposed mechanism facilitates implementing hooks for other components, as described in the example, but said mechanism is not, IMHO, a hook.

Reply:
But do you have issues of using ExecutionHook as API name? What we want to say here is to propose a set of APIs to allow users trigger commands inside their containers